benchexec/tools/klee.py

```diff
@@ -45,30 +45,32 @@ def version(self, executable):
         version = self._version_from_tool(executable, line_prefix="KLEE")
         return version.split("(")[0].strip()
 
-    def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}):
-        if propertyfile:
-            options += ["--property-file=" + propertyfile]
-        if benchexec.model.MEMLIMIT in rlimits:
-            options += ["--max-memory=" + str(rlimits[benchexec.model.MEMLIMIT])]
-        if benchexec.model.TIMELIMIT in rlimits:
-            options += ["--max-time=" + str(rlimits[benchexec.model.TIMELIMIT])]
-        if benchexec.model.WALLTIMELIMIT in rlimits:
-            options += ["--max-walltime=" + str(rlimits[benchexec.model.WALLTIMELIMIT])]
-        if benchexec.model.SOFTTIMELIMIT in rlimits:
-            options += [
-                "--max-cputime-soft=" + str(rlimits[benchexec.model.SOFTTIMELIMIT])
-            ]
-        if benchexec.model.HARDTIMELIMIT in rlimits:
-            options += [
-                "--max-cputime-hard=" + str(rlimits[benchexec.model.HARDTIMELIMIT])
-            ]
+    def cmdline(self, executable, options, task, rlimits):
+        if task.property_file:
+            options += ["--property-file=" + task.property_file]
+        if rlimits.memory:
+            options += ["--max-memory=" + str(rlimits.memory)]
+        if rlimits.cputime_hard:
+            options += ["--max-time=" + str(rlimits.cputime_hard)]
+        if rlimits.walltime:
+            options += ["--max-walltime=" + str(rlimits.walltime)]
+        if rlimits.cputime:
+            options += ["--max-cputime-soft=" + str(rlimits.cputime)]
+        if rlimits.cputime_hard:
+            options += ["--max-cputime-hard=" + str(rlimits.cputime_hard)]
```

Comment:
This changes the semantics, because the old code had an unnoticed bug. `HARDTIMELIMIT` was never present in `rlimits`, so `--max-cputime-hard` was never passed. Now it would get added and always have the same value as `--max-cputime`.

@MartinNowack What is the intention here? BenchExec only has 2 time limits (the actual one and the hard time limit), so 2 parameters would seem to be enough? In which should the main time limit be passed?

Reply:
I had a look at the `klee` wrapper submitted to sv-comp. It seems we can live with this, or even remove. Wouldn't matter much. 
This is how it is used:

```python
    time = 0
    if args.max_cputime_soft: # Use soft timeout if available
        time = args.max_cputime_soft
    elif args.max_cputime_hard:
        time = args.max_cputime_hard
    elif args.max_walltime:
        time = args.max_walltime
    elif args.max_time:
        time = args.max_time
```

Reply:
Good idea for looking at the script!
So they only have one time-limit internally. This should then be set to BenchExec's main time limit (previously named "soft"), not the "hard" time limit, and actually this is also what happened with the previous code.

For the new code, we could simplify this completely to
```
        if rlimits.cputime:
            options += ["--max-time=" + str(rlimits.cputime_hard)]
```
`rlimits.cputime` will always be set if there is some time limit in BenchExec, and it will be set to the value that is relevant for the tool. Passing the other values is irrelevant if the script does not use them anyway I would say.

(Of course, `--max-memory` should be kept.)

@MartinNowack Do you agree?

Reply:
I would say
```
        if rlimits.cputime:
            options += ["--max-cputime-soft=" + str(rlimits.cputime)]
```
because this is also what has the higher priority in the wrapper code for klee.

Reply:
@PhilippWendler wouldn't BenchExec anyways kill the process at hard time limit?

Reply:
Would also be fine for me, @MartinNowack would be best to decide this. If he does not respond, we can still do this.

Reply:
> wouldn't BenchExec anyways kill the process at hard time limit?

Yes, this is what happens.

Reply:
fixed in aaeae7166b8469eebf828b91ee9f6fe768e1afc7

if @MartinNowack prefers some other flag, it can be changed later.